Update Package and Module.md with new changes (#62)
ballerina-platform/module-ballerinax-persist.sql: ballerina/Module.md

# Overview

> This module provides relational database support for the `bal persist` feature, which provides functionality to store and query data from a relational database through a data model instead of writing SQL.

Reviewer: ```suggestion
This package provides relational database support for the `bal persist` feature, which provides functionality to store and query data from a relational database through a data model instead of writing SQL.
```

> Currently, this package supports MySQL, MSSQL and PostgreSQL databases. However, we are also planning to add support for other relational databases such as Oracle.

Reviewer 2: ```suggestion
Currently, this package supports MySQL, MSSQL, and PostgreSQL databases. However, we are also planning to add support for other relational databases such as Oracle.
```

## How to use with `bal persist`

> Since none of the relational databases is not the default datastore for `bal persist`, you need to explicitly specify the data store in your application in either of the following ways,

Reviewer 3: Isn't this should be
```suggestion
Since none of the relational databases is set as the default datastore for `bal persist`, you need to explicitly specify the data store in your application in either of the following ways,
```
?

Author: In the `bal persist add` command, the default datastore is `inmemory` while in `bal persist generate` command, the datastore field is mandotory. Since `inmemory` is not a relational database, this statement is valid. Shall I go ahead and make the change?

Reviewer 3: In that case, something like below would be more clear IMO. WDYT?

```
By default, `bal persist` utilizes the in-memory datastore. Therefore, if you intend to use a different datastore, you must explicitly specify it as follows:
```

Author: +1. Since this package is `persist.sql`, the need to explicitly specify the data store is implied. Hence, what do you think about this?

```
By default, `bal persist` utilizes the in-memory data store. Therefore, you must explicitly specify an SQL data store as follows:
```

Reviewer 3: +1

|     boolean      |       BOOLEAN        |
|      byte[]      |        BYTEA         |
|        ()        |         NULL         |
|    time:Date     |         DATE         |
|  time:TimeOfDay  |         TIME         |
|     time:Utc     |      TIMESTAMP       |
|    time:Civil    |      TIMESTAMP       |
|       enum       | VARCHAR with checks. |

The default length for some SQL types can be changed using the [Advanced SQL type annotations](#mapping-types). If you want even more control over the SQL types, you can change the SQL types in the `script.sql` file generated by the `bal build` or the `persist generate` command before executing the script. It may not work for all the types.

## Advanced SQL annotations

To have custom name and type mappings in the database implementation and to declare indexes, generated fields and custom foreign keys, the below annotations can be used in the data model definition. Note that these annotations can only be used with SQL data stores.

In order to use them, you must first import the `persist.sql` package to your data model definition file as follows.

```ballerina
import ballerinax/persist.sql;
```

> ### Mapping names with `Name` annotation

Reviewer 2: ```suggestion
### Map name with `Name` annotation
```

> - Map entity names to table names

Reviewer 2: ```suggestion
- Map entity name to table name
```

```ballerina
@sql:Name {value: "people"}
type Person record {|
    readonly int id;
    string name;
    string address;
|};
```

The `Person` entity will be mapped to the `people` table in the database.

> - Map field names to column names

Reviewer 2: ```suggestion
- Map field name to column name
```

```ballerina
type Person record {|
    @sql:Name {value: "person_id"}
    readonly int id;
    @sql:Name {value: "full_name"}
    string name;
    string address;
|};
```

The `id` field will be mapped to the `person_id` column in the database, and the `name` field will be mapped to the `full_name` column in the database.

> ### Mapping types

Reviewer 2: ```suggestion
### Type Mapping
```

#### `Varchar` annotation

```ballerina
type Person record {|
    readonly int id;
    @sql:Varchar {length: 100}
    string name;
    string address;
|};
```

The `name` field will have a `VARCHAR(100)` column in the database. The `@sql:Varchar` annotation can only be used on `string` fields.

#### `Char` annotation

```ballerina
type Person record {|
    @sql:Char {length: 12}
    readonly string nic;
    string name;
    string address;
|};
```

The `nic` field will have a `CHAR(12)` column in the database. The `@sql:Char` annotation can only be used on `string` fields.

#### `Decimal` annotation

```ballerina
type Person record {|
    readonly int id;
    @sql:Decimal {precision: [10,2]}
    decimal salary;
    string address;
|};
```

The `salary` field will have a `DECIMAL(10,2)` column in the database. The `@sql:Decimal` annotation can only be used on `decimal` fields.

> ### Declaring indexes

Reviewer 2: ```suggestion
### Declare indexes
```

#### `UniqueIndex` annotation

This annotation lets you define a unique index on a specific field, and the usage is similar to the `@sql:Index` annotation.

```ballerina
type Person record {|
    readonly int id;
    @sql:UniqueIndex {name: "idx_person"}
    string nic;
    @sql:UniqueIndex {name: ["idx_person", "idx_another"]}
    string name;
    @sql:UniqueIndex
    string address;
|};
```

The unique index `idx_person` is a composite unique index consisting of `nic` and `name` fields. `idx_another` is just another unique index on the `name` field. The `address` field also has the `@sql:UniqueIndex` annotation without the `name` property. Here, the index name will be generated by `persist` in the `unique_idx_[FIELD_NAME]` format, in which case the index name for the `address` field becomes `unique_idx_address`.

> ### Declaring generated fields with `Generated` annotation

Reviewer 2: ```suggestion
### Declare generated fields with `Generated` annotation
```

The `@sql:Generated` annotation is used to declare a field as a generated field. This annotation can only be used on `readonly int` fields. Currently, only the `AUTO_INCREMENT` or an equivalent generation strategy is supported.

```ballerina
type Person record {|
    @sql:Generated
    readonly int id;
    string name;
    string address;
|};
```

The `id` field will be auto-generated and the [`PersonInsert`](/learn/persist-client-api/#insert-types) type will not have the `id` field as it should be auto-generated by the database. Refer the below table to understand how the generation strategy is implemented in different data stores.

|  Data store  |  Implementation  |
|:------------:|:----------------:|
|   `MySQL`    | `AUTO_INCREMENT` |
|   `MSSQL`    | `IDENTITY(1,1)`  |
| `PostgreSQL` |     `SERIAL`     |

> ### Declaring custom foreign keys with `Relation` annotation

Reviewer 2: ```suggestion
### Declare custom foreign keys with `Relation` annotation
```

> The `@sql:Relation` annotation can be used to declare your own custom foreign key field name. You must put the foreign key on the correct side of the relationship (ownerside) and the key field must exist on the record type, and must be of the same type as the primary key of the referred entity. This is particularly useful when a foreign key is also a part of the composite primary key.

Reviewer 2: ```suggestion
The `@sql:Relation` annotation can be used to declare your own custom foreign key field. You must put the foreign key on the correct side of the relationship (owner side) and the key field must exist on the record type and must be of the same type as the primary key of the referred entity. This is particularly useful when a foreign key is also a part of the composite primary key.
```

```ballerina
type Car record {|
   readonly int id;
   string name;
   int userId;
   @sql:Relation {keys: ["userId"]}
   User owner;
|};

type User record {|
   readonly int id;
   string name;
   Car[] cars;
|};
```

> The `keys` field accepts an array of foreign keys and the length of it must be same as the number of primary keys the referring entity has. Here, the field `userId` is used as the foreign key for the relation `owner` and it has been declared so through the `@sql:Relation` annotation. When the `@sql:Relation` annotation is used in a relation field, the foreign key will not be generated by default.

Reviewer 2: ```suggestion
The `keys` field accepts an array of foreign keys and the length of it must be the same as the number of primary keys the referring entity has. Here, the field `userId` is used as the foreign key for the relation `owner` and it has been declared so through the `@sql:Relation` annotation. When the `@sql:Relation` annotation is used in a relation field, the foreign key will not be generated by default.
```

|  password  |    The password of the DB server.    |
|  database  | The name of the database to be used. |

> The following is a sample `Config.toml` file with the MySQL data store configuration. This will be generated by the `bal build` or `bal persist generate` command (depending on how you have configured the application).

Reviewer 2: ```suggestion
The following is a sample `Config.toml` file with the MySQL data store configuration. This will be generated by the `bal build` or `bal persist generate` command (depending on how you generate the client API).
```
